### PR TITLE
docs: add doc comparing jj to Sapling (#1708)

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: docs/related-work.md [hunk 1]
```diff
@@ -9,7 +9,8 @@ Similar tools:
 * [Sapling](https://sapling-scm.com/): A heavily modified fork of
   [Mercurial](https://www.mercurial-scm.org/) developed and used at Meta. It
   is compatible with Git, has undo functionality, and
-  [a graphical interface](https://sapling-scm.com/docs/addons/isl). 
+  [a graphical interface](https://sapling-scm.com/docs/addons/isl).
+  See [how it is difference from Jujutsu](sapling-comparison.md).
```
Comment (PhilipMetzger) [inline docs/related-work.md:13]: Typo: `how it is difference from Jujutsu` => `how it is different from Jujutsu` or `See the difference to Jujutsu`.

**Review by PhilipMetzger — APPROVED** [review]
LGTM, but it would be valuable if a Meta employee also took a look at it.

**Comment by quark-zju** [comment]
Mostly look good to me. There are some future plans that might count as noticeable differences but I think it's fine not documenting them until they are available:
- Sapling might integrate with the virtual filesystem, EdenFS, and indirectly watchman and buck2 (i.e. buck2 can check file content hashes without materializing them on the filesystem). That might justify why it does not co-exist in a same git checkout (since git does not integrate with EdenFS).
- The underlying storage is different. Sapling mainly uses IndexedLog (and its variants RotateLog, and ZStore) so `.sl` inode count is bounded. As a trade-off Sapling uses file locking and might not work well with NFS (although EdenFS technically can expose its Thrift RPC over the network for remote use-cases). It's a different trade-off from jj. Although right now it's mostly libgit2 and Sapling does not yet use its internal storage library yet.

File: docs/sapling-comparison.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+* **Undo:** jj's undo is powered by [the operation log](operation-log.md), which
+  records how the repo has changed over time. Sapling has a similar feature
+  with its [MetaLog](https://sapling-scm.com/docs/internals/metalog).
+  They seem to provide similar functionality, but jj also exposes the log to the
+  user via `jj op log`, so you can tell how far back you want to go back.
+  Sapling has `sl debugmetalog`, but that seems to show the history of a single
+  commit, not the whole repo's history. Thanks to jj snapshotting the working
+  copy, it's possible to undo changes to the working copy. For example, if
+  you `jj undo` a ` jj commit`, `jj diff` will show the same changes as
+  before `jj commit`, but if you `sl undo` a `sl commit`, the working copy will
```
Comment (zzl0) [inline docs/sapling-comparison.md:53]: minor: you can use `sl undo -k` to keep working copy changes

File: docs/related-work.md (hunk 1 above)
Comment (martinvonz) [reply docs/related-work.md:13]: Thanks, fixed

File: docs/sapling-comparison.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  Git repo. jj also does, and it also supports sharing a working copy with a Git
+  repo, so you can use `jj` and `git` interchangeably in the same repo.
+* **Polish:** Sapling is much more polished and feature-complete. For example,
+  jj has no `blame/annotate` or `bisect` commands, and also no copy/rename
+  support. Sapling also has very nice web UI
+  called [Interactive Smartlog](https://sapling-scm.com/docs/addons/isl), which
+  lets you drag and drop commits to rebase them, among other things.
+* **Forge integrations:** Sapling has `sl pr submit --stack`, which lets you
+  push a stack of commits as separate GitHub PRs (or GitLab MRs, since it simply
+  creates branches?). You will then have to manually add the base branch for
```
Comment (zzl0) [inline docs/sapling-comparison.md:65]: > You will then have to manually add the base branch for each PR in GitHub's UI

Sorry, I don't follow one. Do you mean users need to do this when merging the PRs or creating PRs?

**Review by zzl0 — COMMENTED** [review]
LGTM. Just two minor comments.

File: docs/sapling-comparison.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    * Rebasing descendants always succeeds. Like jj, Sapling automatically
+      rebases, but it will fail if there are conflicts.
+    * Merge commits can be rebased correctly (Sapling sometimes fails).
+    * You can rebase conflicts and conflict resolutions.
+* **Undo:** jj's undo is powered by [the operation log](operation-log.md), which
+  records how the repo has changed over time. Sapling has a similar feature
+  with its [MetaLog](https://sapling-scm.com/docs/internals/metalog).
+  They seem to provide similar functionality, but jj also exposes the log to the
+  user via `jj op log`, so you can tell how far back you want to go back.
+  Sapling has `sl debugmetalog`, but that seems to show the history of a single
```
Comment (strager) [inline docs/sapling-comparison.md:49]: Is `sl journal` the command you're looking for? Sample output:
```
previous locations of '.':
24ee3b26e7df -> ddcd52e862d0 strager  2023-06-09 01:36 -0700  goto -- master
5c4c4226f5c8 -> 24ee3b26e7df strager  2023-06-07 00:47 -0700  goto -- master
37166b2607e7 -> 5c4c4226f5c8 strager  2023-06-04 22:05 -0700  split
5c4c4226f5c8 -> 37166b2607e7 strager  2023-06-04 22:02 -0700  split
c8c06749ace3 -> 5c4c4226f5c8 strager  2023-06-03 01:36 -0700  goto -- master
e5b36311f1cb -> c8c06749ace3 strager  2023-06-02 01:44 -0700  commit -v
7cb6ac7b737e -> e5b36311f1cb strager  2023-06-02 01:43 -0700  commit -v
ce96893d1ba9 -> 7cb6ac7b737e strager  2023-06-02 00:57 -0700  goto -- master
```

File: docs/sapling-comparison.md (hunk 3 above)
Comment (strager) [inline docs/sapling-comparison.md:65]: `sl pr` uses GitHub's API thus does not work with GitLab. (`sl pr submit` does not just create branches.)

**Comment by martinvonz** [comment]
> Mostly look good to me. There are some future plans that might count as noticeable differences but I think it's fine not documenting them until they are available:

Thanks for reviewing! Yes, I think it's better to document that once it's done. Feel free to send a PR when it is.

> * Sapling might integrate with the virtual filesystem, EdenFS, and indirectly watchman and buck2 (i.e. buck2 can check file content hashes without materializing them on the filesystem). That might justify why it does not co-exist in a same git checkout (since git does not integrate with EdenFS).

That actually sounds similar to what we're thinking of doing with jj at Google. I hope we'll be able to do that by replacing the working copy implementation (it's not currently a trait, but it will be soon). So we'll have the default working-copy implementation, which is used outside of Google, and we'll have a custom CitC-aware implementation internally. The right working-copy implementation will be loaded depending on the repo, just like we load the right commit backend etc. So maybe the main difference will be that your VFS-integration will be open-sourced.

> * The underlying storage is different. Sapling mainly uses IndexedLog (and its variants RotateLog, and ZStore) so `.sl` inode count is bounded. As a trade-off Sapling uses file locking and might not work well with NFS (although EdenFS technically can expose its Thrift RPC over the network for remote use-cases). It's a different trade-off from jj. Although right now it's mostly libgit2 and Sapling does not yet use its internal storage library yet.

Ah, that's a good point. I think it's probably too rare that users care about NFS support that it's worth mentioning.

File: docs/sapling-comparison.md (hunk 3 above)
Comment (martinvonz) [reply docs/sapling-comparison.md]: > Sorry, I don't follow this line. Do you mean users need to do this when merging the PRs or creating PRs?

When creating PRs.

> Not sure if you are aware of the various workflows supported by `sl pr submit`.
> 
> https://github.com/facebook/sapling/blob/a66a40a339e939aa2000d849137799edd0cbddc4/eden/scm/edenscm/ext/github/submit.py#L39-L65

Thanks, I wasn't at all aware (I have never tried `sl pr submit`, so what I wrote here was just based on what I found on https://sapling-scm.com/docs/git/sapling-stack. That's nice that `sl pr submit` updates the base branches! I've updated the text here.

Comment (martinvonz) [reply docs/sapling-comparison.md]: Makes sense. Is there a way to just create branches on an non-GitHub remote? For now, I've updated the text to say that only GitHub is supported, but I'm happy to correct it if Sapling supports pushing branches. Maybe `sl push` can do that?

Comment (strager) [reply docs/sapling-comparison.md]: > Is there a way to just create branches on an non-GitHub remote?

Yes, I do this all the time (with GitHub). `sl push -r . --to master` pushes the current commit (`.`) to a remote branch called `master`, for example.

Sapling can also manage remote branches. For example, `sl push --delete tempbranch` behaves like `git push origin :refs/heads/tempbranch`.

> For now, I've updated the text to say that only GitHub is supported

I thought this section is about *forge-specific integrations*, not general Git server support. Sapling does support pushing to and pulling from Git servers.

**Comment by strager** [comment]
If jj supports Git tags, you could mention that Sapling does *not* support Git tags.

File: docs/sapling-comparison.md (hunk 3 above)
Comment (zzl0) [reply docs/sapling-comparison.md]: Yeah, we should update the document of `sl pr submit`.

Comment (martinvonz) [reply docs/sapling-comparison.md]: > I thought this section is about _forge-specific integrations_, not general Git server support. Sapling does support pushing to and pulling from Git servers.

Ah, good point, "integrations" wasn't correct for what I meant. I've changed it to "Forge workflows".

**Comment by martinvonz** [comment]
> If jj supports Git tags, you could mention that Sapling does _not_ support Git tags.

Thanks, but jj has very limited support (you can see and specify revisions by tag name, but you can't create tags), so I don't think it's worth mentioning.

**Pull request merged**